gto/git_utils.py

```diff
@@ -0,0 +1,71 @@
+import inspect
+import logging
+from contextlib import contextmanager
+from functools import wraps
+from tempfile import TemporaryDirectory
+from typing import Callable, Dict
+
+from git import Repo
+
+from gto.constants import remote_git_repo_regex
+
+
+def git_clone_if_repo_is_remote(f: Callable):
+    @wraps(f)
+    def wrapped_f(*args, **kwargs):
+        kwargs = _turn_args_into_kwargs(args, kwargs)
+
+        # noinspection PyTypeChecker
+        if isinstance(kwargs["repo"], str) and is_url_of_remote_repo(
+            repo=kwargs["repo"]
+        ):
+            try:
+                # noinspection PyTypeChecker
+                with cloned_git_repo(repo=kwargs["repo"]) as tmp_dir:
+                    kwargs["repo"] = tmp_dir
+                    return f(**kwargs)
+            except (NotADirectoryError, PermissionError) as e:
+                raise e.__class__(
+                    "Are you using windows with python < 3.9? "
+                    "This may be the reason of this error: https://bugs.python.org/issue42796. "
+                    "Consider upgrading python."
+                ) from e
+
+        return f(**kwargs)
+
+    def _turn_args_into_kwargs(
+        args: tuple, kwargs: Dict[str, object]
+    ) -> Dict[str, object]:
+        kwargs_complement = {
+            k: args[i]
+            for i, k in enumerate(inspect.getfullargspec(f).args)
+            if i < len(args)
+        }
+        kwargs.update(kwargs_complement)
+        return kwargs
+
+    return wrapped_f
+
+
+def is_url_of_remote_repo(repo: str) -> bool:
+    if remote_git_repo_regex.fullmatch(repo) is not None:
+        logging.debug("%s recognized as remote git repo", repo)
+        return True
+
+    logging.debug("%s NOT recognized as remote git repo", repo)
+    return False
+
+
+@contextmanager
+def cloned_git_repo(repo: str):
+    tmp_dir = TemporaryDirectory()
+    logging.debug("create temporary directory %s", tmp_dir)
+    git_clone(repo=repo, dir=tmp_dir.name)
+    yield tmp_dir.name
+    logging.debug("delete temporary directory %s", tmp_dir)
+    tmp_dir.cleanup()
+
+
+def git_clone(repo: str, dir: str) -> None:
+    logging.debug("clone %s in directory %s", repo, dir)
+    Repo.clone_from(url=repo, to_path=dir)
```

Comment:
Let's merge these functions and test a single one only? Now having `git_clone` as a separate seems redundant. You can easily test `cloned_git_repo` without generating a temporary directory additionally in tests.

Reply:
Can gladly do it for the tests (actually it connects with something I wanted to discuss about public/private functions -> will add it below), but for the functions I would suggest to leave them as they are, again to leave the door open for a different engine. It doesn't do any harm and makes the separation clear: this is the only thing you need to change if you want to use a different git engine.

Reply:
Thanks, I checked out the second PR you created.

My opinion:
1. We should keep tests. We can remove them anytime later. Keeping the functions public is also OK for me, this is in the fashion of the rest of GTO codebase I think.
2. We can keep these functions separate, doesn't matter for now I guess.

I think the PR is ready, no need to overthink this. We can get fix all of that when that would be needed. Let's merge this and move along. WDYT?

Reply:
fine with me! :) let's do that